File: src/lowlevel/server/icefsoperators/icefsUnlink.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 17:20:03
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsUnlink.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"

	"golang.org/x/sys/unix"
)

type IcefsDoUnlinkGetNlinkInode func(entry any) (stNlink uint64, inode uint64)

func GRpcIcefsDoUnlinkGetNlinkInode(entry any) (stNlink uint64, inode uint64) {
	return entry.(*pb.FuseEntryParam).Attr.StNlink, entry.(*pb.FuseEntryParam).Inode
}

func ThriftIcefsDoUnlinkGetNlinkInode(entry any) (stNlink uint64, inode uint64) {
	return uint64(entry.(*icefsthrift.FuseEntryParam).Attr.StNlink), uint64(entry.(*icefsthrift.FuseEntryParam).Inode)
}

func (s *IcefsServer) doIcefsUnlink(parentFakeInode uint64, name string, fuseEntryParamBuilder FuseEntryParamBuilder, icefsDoUnlinkGetNlinkInode IcefsDoUnlinkGetNlinkInode) (status int32) {
	var err error
	var parentInode *IcefsInode
	if s.timeout == 0 {
		entry, err := s.doIcefsLookUp(parentFakeInode, name, fuseEntryParamBuilder)
		if err != nil {
			status = int32(syscall.EIO)
		}
		stNlink, fakeInode := icefsDoUnlinkGetNlinkInode(entry)
		if stNlink == 1 {
			s.inodeCacheLock.RLock()
			inode := s.getIcefsInode(fakeInode)
			if inode == nil {
				s.inodeCacheLock.RUnlock()
				status = icefserror.ICEFS_BUG_ERR
				return
			}
			inode.inodeLock.Lock()
			s.inodeCacheLock.RUnlock()
			if inode.fd > 0 && inode.nopen == 0 {
				syscall.Close(inode.fd)
				inode.fd = -int(syscall.ENOENT)
				inode.generation++
			}
			inode.inodeLock.Unlock()
		}
		s.doIcefsForget(fakeInode, 1)
	}

	s.inodeCacheLock.RLock()
	parentInode = s.getIcefsInode(parentFakeInode)
	if parentInode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}

	parentInode.inodeLock.RLock()
	s.inodeCacheLock.RUnlock()
	err = unix.Unlinkat(parentInode.fd, name, 0)
	parentInode.inodeLock.RUnlock()

	status = icefserror.IcefsStdErrno(err)
	return
}

func (s *IcefsGRpcServer) DoIcefsUnlink(ctx context.Context, req *pb.IcefsUnlinkReq) (*pb.IcefsUnlinkRes, error) {
	var res pb.IcefsUnlinkRes

	res.Status = s.server.doIcefsUnlink(req.ParentInode, req.Name, GRpcFuseEntryParamBuilder, GRpcIcefsDoUnlinkGetNlinkInode)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsUnlink(ctx context.Context, req *icefsthrift.IcefsUnlinkReq) (*icefsthrift.IcefsUnlinkRes, error) {
	var res icefsthrift.IcefsUnlinkRes

	res.Status = s.server.doIcefsUnlink(uint64(req.ParentInode), req.Name, ThriftFuseEntryParamBuilder, ThriftIcefsDoUnlinkGetNlinkInode)

	return &res, nil
}

File: src/lowlevel/client/icefsoperators/icefsUnlink.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:41:07
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsUnlink.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsUnlink(fuse_req_t fuseReq, fuse_ino_t parentInode,
                                const char *name) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsUnlinkReq req;
      icefsgrpc::IcefsUnlinkRes res;
      grpc::ClientContext ctx;

      req.set_parent_inode(parentInode);
      req.set_name(name);

      grpc::Status status = gRpcClient->DoIcefsUnlink(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        fuse_reply_err(fuseReq, ICEFS_EOK);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsUnlinkReq req;
      icefsthrift::IcefsUnlinkRes res;

      req.__set_parent_inode(parentInode);
      req.__set_name(name);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsUnlink(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      fuse_reply_err(fuseReq, res.status);
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}